File: process/constructions/column_increase_1.pl
#!/usr/bin/perl

my $name = 'column_increase_1';
my $id;

$id = get_construction_id( $name, 1, 2 );

$constructions{ $name } = {};

$constructions{ $name }->{'process_array'} =
	sub {
		my $array = shift;

		return if ( $array->{'k'} >= $MAXK );

		# first, we use this array as the main array

		if ( $array->{'k'} >= $array->{'N'} ) {
			if ( $array->{'t'} > 3 ) {

				my $subarray = get_best_array( $array->{'k'}, $array->{'v'} - 2, $array->{'t'} - 2 );

				if ( $subarray ) {
					my $new_array = new_constructed_array( $array->{'N'} + $subarray->{'N'},
					                                       $array->{'k'} + 1,
						   								   $array->{'v'},
														   $array->{'t'} );

					$new_array->{'construction'} = $id;
					$new_array->{'ingredients'} = [ $array, $subarray ];

					if ( $new_array->{'k'} >= $new_array->{'N'} ) {
						persist_constructed_array( $new_array );
					}
				}
			}
		}

		# we also need to try this array as the smaller array

		if ( ($array->{'t'} + 2) > $MAXT || ($array->{'v'} + 2 > $MAXV) ) {
			return;
		}

		my $prev = get_max_k_array( 'k < ' . $array->{'k'} . ' AND t = ' . $array->{'t'} . ' AND v = ' . $array->{'v'} );

		my $condition = 'k <= ' . $array->{'k'} . ' AND t = ' . ($array->{'t'}+2) . ' AND v = ' . ($array->{'v'}+2);
		if ( $prev ) {
			$condition .= ' AND k > ' . $prev->{'k'};
		}
		my $arrays = get_arrays( $condition );
		my $superarray;
		foreach $superarray (@$arrays) {
			my $new_array = new_constructed_array( $array->{'N'} + $superarray->{'N'},
		                                       	   $superarray->{'k'} + 1,
			   								       $array->{'v'} + 2,
											       $array->{'t'} + 2 );
			$new_array->{'construction'} = $id;
			$new_array->{'ingredients'} = [ $superarray, $array ];

			if ( $new_array->{'k'} >= $new_array->{'N'} ) {
				persist_constructed_array( $new_array );
			}
		}
	};

1;

File: process/datastore/constructions/column_increase_1.cpp
#include "construction.h"
#include "../datastore.h"
#include "../util/math.h"

class c_column_increase_1 : public construction {
public:
	c_column_increase_1() : construction( "column_increase_1", 13 ) {
	}

	void process( array * base ) {
		if ( base->k >= MAXK ) {
			return;
		}

		// first, we use this array as the main array

		if ( base->t > 3 ) {
			array * subarray = getBestArray( base->k, base->v - 2, base->t - 2 );

			if ( subarray != NULL ) {
				array * arr = new array();
				arr->N = base->N + subarray->N;
				arr->k = base->k + 1;
				arr->v = base->v;
				arr->t = base->t;

				arr->type = 'C';
				arr->source = id;

				arr->ingredients.push_back( base );
				arr->ingredients.push_back( subarray );

				insertArray( arr );
			}
		}

		// we also need to try this array as the smaller array

		if ( ( base->t + 2) > MAXT || (base->v + 2 > MAXV) ) {
			return;
		}

		array * prev = getPreviousArray( base );

		fullSortTableType arrays;
		query( arrays, base->t + 2, base->t + 2, base->v + 2, base->v + 2, prev != NULL ? prev->k + 1 : 0, base->k );

		fullSortTableType::const_iterator superIt;
		for ( superIt = arrays.begin(); superIt != arrays.end(); superIt++ ) {
			array * superarray = *superIt;

			array * arr = new array();
			arr->N = base->N + superarray->N;
			arr->k = superarray->k + 1;
			arr->v = superarray->v;
			arr->t = superarray->t;

			arr->type = 'C';
			arr->source = id;
			
			arr->ingredients.push_back( superarray );
			arr->ingredients.push_back( base );

			insertArray( arr, &arrays );
		}
	}
};

// declare the construction object
construction * column_increase_1 = new c_column_increase_1();